### PR TITLE
Adding datepicker double input + @Output

### PR DESCRIPTION
**Pull request by VicoDraye**

### CODE REVIEW

File: src/app/datepicker/datepicker.component.ts [hunk 1]
```diff
@@ -1,32 +1,108 @@
...
+  
+  datepickerRange = {
+    begin: new Date(),
+    end: new Date(),
+  };
+
+  rangeFormGroup: FormGroup;
+  isValidRange= false;
+
+  constructor(private fb:FormBuilder){
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: Don't use FormBuilder, use `FormGroup` directly instead. See below.

File: src/app/datepicker/datepicker.component.ts [hunk 2]
```diff
@@ -1,32 +1,108 @@
...
+  }
+
+  ngOnInit(): void {
+   this.rangeFormGroup = this.fb.group(
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: Validators should be pure functions (outside of the class): 
```typescript
this.rangeForm = new FormGroup({
  startRange: new FormControl(new Date(), startDateValidator),
  endRange: new FormControl(new Date(), endDateValidator)
})
```

File: src/app/datepicker/datepicker.component.ts [hunk 3]
```diff
@@ -1,32 +1,108 @@
...
+    {
+      startRange:[new Date(),[this.startDateValidator]],
+      endRange:[new Date(),[this.endDateValidator]],  
+    },);
+  }
+    
+  public addRange(){
+    this.wantedRange.emit({startRange: this.datepickerRange.begin, endRange: this.datepickerRange.end});
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: For readability : 
```typescript
const { begin, end } = this.datepickerRange;
this.wantedRand.emit({ startRange: begin, endRange: end });
```
But I would recommand to match the interface from datepickerRange. Something like that : 
```typescript
const { begin, end } = this.datepickerRange;
this.wantedRand.emit({ begin, end });
```

File: src/app/datepicker/datepicker.component.ts [hunk 4]
```diff
@@ -1,32 +1,108 @@
...
-  }
+  }    
 
-  /** disable all dates who are between the periods in disabledDates[] */
-  disableDates = (d: Date): boolean => {
-    for (let disableRange of this.disabledDates){
-      if (isBetween(d, disableRange.startRange, disableRange.endRange)){
-        return false;
-      }
+ disableDates = (d: Date): boolean => {
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: Do :
```typescript
public disableDates(date: Date): boolean {
  return !this.disabledDates.some(({ begin, end }) => isBetween(date, begin, end));
}
```

File: src/app/datepicker/datepicker.component.ts [hunk 5]
```diff
@@ -1,32 +1,108 @@
...
+  for (let disableRange of this.disabledDates){
+    if (isBetween(d, disableRange.startRange, disableRange.endRange)){
+      return false;
     }
+  }
     return true;
   }
+
+  startRangeChange(value) {
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: type the value : `startRangeChanged(value: number)` for example.

Comment (GrandSchtroumpf) [reply src/app/datepicker/datepicker.component.ts]: Do the same for ALL parameters of ALL methods

File: src/app/datepicker/datepicker.component.ts [hunk 6]
```diff
@@ -1,32 +1,108 @@
...
   styleUrls: ['./datepicker.component.scss'],
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class DatepickerComponent {
+export class DatepickerComponent implements OnInit {
   
   @Input() selectedRange: CalendarRange;
   @Input() disabledDates: CalendarRange[];
 
+  @Output() wantedRange = new EventEmitter(); ;
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: Type your EventEmitter: `new EventEmitter<{ begin: number, end: number}>()` for example. Ideally create an interface for the type.

File: src/app/datepicker/datepicker.component.ts [hunk 7]
```diff
@@ -1,32 +1,108 @@
...
 import { MatDatepicker } from '@angular/material';
+
 import { CalendarRange } from '../calendar-view/calendar-view.component';
-import { isBetween } from '../utils'
+import { isBetween } from '../utils';
+import { FormControl, FormGroup, FormBuilder } from '@angular/forms';
+
 
 @Component({
   selector: 'app-datepicker',
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: Give another name for that. Like "datepicker-range"

File: src/app/datepicker/datepicker.component.ts [hunk 8]
```diff
@@ -1,32 +1,108 @@
...
+    }
+  }
+
+  datepickerRangeChange($event) {
+    this.datepickerRange = $event.value;
+    this.validateDapickerInput(this.datepickerRange);
+  }
+
+  validateDapickerInput(datepickerRange){
+    if(!!datepickerRange.begin && !!datepickerRange.end){
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: Do the opposite : 
```typescript
const { begin, end } = datepicker;
if (!begin || !end) {
  return; // or throw new Error()
}
const isInvalid = (disabledDate) => {
  return isBetween(disabledDate.begin, begin, end) || isBetween(disabledDate.end, begin, end);
}
this.isValidRange = !this.disabledDates.some(disabledDate => isInvalid(disabledDate));
return this.isValidRange;
```

File: src/app/datepicker/datepicker.component.html [hunk 9]
```diff
@@ -1,21 +1,21 @@
-<form [formGroup]="rangeFormGroup">
+<form [formGroup]="rangeForm">
   <div>
     <mat-form-field appearance="outline">
       <mat-label>Start date</mat-label>
-      <input #startRange matInput placeholder="Start date" [value]="datepickerRange.begin | date:'yyyy-MM-dd'" formControlName="startRange" (blur)="startRangeChange(startRange.value)">
-      <mat-error *ngIf="rangeFormGroup.get('startRange').hasError('required')">
+      <input #begin matInput placeholder="Start date" [value]="datepickerRange.begin | date:'yyyy-MM-dd'" formControlName="begin" (blur)="startRangeChange(begin.value)">
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.html]: Why do you need a reference to "begin" here ? I don't see where you are using it. Maybe you can remove it.

Comment (GrandSchtroumpf) [reply src/app/datepicker/datepicker.component.html]: Also, remove the "value", it should be managed by the formControlName directly.

File: src/app/datepicker/datepicker.component.html [hunk 10]
```diff
@@ -1,21 +1,21 @@
...
       </mat-error>
-      <mat-error *ngIf="rangeFormGroup.get('startRange').hasError('wrongStartRange')">Please provide a valid start date</mat-error>  
+      <mat-error *ngIf="rangeForm.get('begin').hasError('wrongStartRange')">Please provide a valid start date</mat-error>  
     </mat-form-field>
     -
     <mat-form-field appearance="outline">
       <mat-label>End date</mat-label>
-      <input #endRange matInput placeholder="End date" [value]="datepickerRange.end | date:'yyyy-MM-dd'" formControlName="endRange" (blur)="endRangeChange(endRange.value)">
-      <mat-error *ngIf="rangeFormGroup.get('endRange').hasError('required')">
+      <input #end matInput placeholder="End date" [value]="datepickerRange.end | date:'yyyy-MM-dd'" formControlName="end" (blur)="endRangeChange(end.value)">
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.html]: Same as above

File: src/app/datepicker/datepicker.component.ts [hunk 11]
```diff
@@ -17,92 +17,77 @@ export class DatepickerComponent implements OnInit {
...
+  disableDates = (date: Date): boolean => {
+    return !this.disabledDates.some(({ begin, end }) => isBetween(date, begin, end));
   }
 
-  startRangeChange(value) {
-    if (!isNaN(Date.parse(value))){
-    this.datepickerRange.begin = new Date(value);
+  startRangeChange(date: string) {
+    if (!isNaN(Date.parse(date))){
+    this.datepickerRange.begin = new Date(date);
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: Add tab here for formating.

File: src/app/datepicker/datepicker.component.ts [hunk 12]
```diff
@@ -17,92 +17,77 @@ export class DatepickerComponent implements OnInit {
...
     this.validateDapickerInput(this.datepickerRange);
     }
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: What happend if this is NaN ?

File: src/app/datepicker/datepicker.component.ts [hunk 13]
```diff
@@ -17,92 +17,77 @@ export class DatepickerComponent implements OnInit {
...
-    this.datepickerRange.end = new Date(value);
+  endRangeChange(date: string) {
+    if (!isNaN(Date.parse(date))){
+    this.datepickerRange.end = new Date(date);
     this.validateDapickerInput(this.datepickerRange);
     }
   }
 
   datepickerRangeChange($event) {
+    console.log($event);
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: Remove that.

File: src/app/datepicker/datepicker.component.ts [hunk 14]
```diff
@@ -17,92 +17,77 @@ export class DatepickerComponent implements OnInit {
...
-    if (!isNaN(Date.parse(value))){
-    this.datepickerRange.end = new Date(value);
+  endRangeChange(date: string) {
+    if (!isNaN(Date.parse(date))){
+    this.datepickerRange.end = new Date(date);
     this.validateDapickerInput(this.datepickerRange);
     }
   }
 
   datepickerRangeChange($event) {
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: Give a name (something else than $event) to your event so we know what it is about. Also type it.

File: src/app/datepicker/datepicker.component.ts [hunk 15]
```diff
@@ -17,92 +17,77 @@ export class DatepickerComponent implements OnInit {
...
-    if(!!datepickerRange.begin && !!datepickerRange.end){
-      this.isValidRange = true;
-      this.disabledDates.forEach(disabledDate => {
-        if(isBetween(disabledDate.startRange, datepickerRange.begin, datepickerRange.end) 
-        || isBetween(disabledDate.endRange, datepickerRange.begin, datepickerRange.end)){
-          this.isValidRange = false;
-        }
-      });
-    return this.isValidRange;
+  validateDapickerInput(datepicker: CalendarRange){
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: Typo in the name

File: src/app/datepicker/datepicker.component.ts [hunk 16]
```diff
@@ -17,92 +17,77 @@ export class DatepickerComponent implements OnInit {
...
+    const { begin, end } = datepicker;
+    if (!begin || !end) {
+      return; // or throw new Error()
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: Please do copy/paste the code that I write on comment, without trying to understand what it does. Either you return nothing (this is not an error), or you throw an error. Or you display something to the user with a MatSnackBar...

File: src/app/datepicker/datepicker.component.ts [hunk 17]
```diff
@@ -17,92 +17,77 @@ export class DatepickerComponent implements OnInit {
   @Input() selectedRange: CalendarRange;
   @Input() disabledDates: CalendarRange[];
 
-  @Output() wantedRange = new EventEmitter(); ;
+  @Output() wantedRange = new EventEmitter<{begin: Date, end: Date}>(); ;
```
Comment (GrandSchtroumpf) [inline src/app/datepicker/datepicker.component.ts]: You have a type of that "CalendarRange" : `EventEmitter<CalendarRange>`.

File: src/app/datepicker/datepicker.component.ts (hunk 12 above)
Comment (VicoDraye) [reply src/app/datepicker/datepicker.component.ts]: If it is NaN, it means that the date entered by the user is wrong

Comment (GrandSchtroumpf) [reply src/app/datepicker/datepicker.component.ts]: Then you want to tell the user that maybe. With a MatSnackBar

File: src/app/datepicker/datepicker.component.html [hunk 18]
```diff
@@ -1,17 +1,25 @@
...
+  <mat-form-field appearance="outline" class="">
+    <mat-label>Start date - End date</mat-label>
     <input 
       matInput
+      [formControl]="range"
       [matDatepickerFilter]="disableDates"
       placeholder="Choose a date"
       [satDatepicker]="picker"
+      (dateChange)="datepickerRangeChange($event.value)"
+      (blur)="datepickerRangeChange($event.value)"
```
Comment (pldespaigne) [inline src/app/datepicker/datepicker.component.html]: `(blur)="datepickerRangeChange($event)"`, retreive `.value` inside the `datepickerRangeChange()` function,
same thing for the line above

File: src/app/datepicker/datepicker.component.ts [hunk 19]
```diff
@@ -14,19 +18,69 @@ export class DatepickerComponent {
...
+  
+  isValidRange= false;
+  range = new FormControl('', [Validators.required]);
+
+
+  constructor(private _snackBar: MatSnackBar) {}
+    
+  public addRange(){
+    const { begin, end } = this.datepickerRange;
+    this.wantedRange.emit({ begin, end });
```
Comment (pldespaigne) [inline src/app/datepicker/datepicker.component.ts]: `this.wantedRange.emit(this.datepickerRange);`

File: src/app/datepicker/datepicker.component.ts [hunk 20]
```diff
@@ -14,19 +18,69 @@ export class DatepickerComponent {
...
+  }
+
   initDatePicker(){
     this.picker.open();
+    }
```
Comment (pldespaigne) [inline src/app/datepicker/datepicker.component.ts]: too much indentation

File: src/app/datepicker/datepicker.component.ts [hunk 21]
```diff
@@ -14,19 +18,69 @@ export class DatepickerComponent {
...
+    const { begin, end } = this.datepickerRange;
+    this.wantedRange.emit({ begin, end });
+  }
+
   initDatePicker(){
     this.picker.open();
+    }    
+
+  disableDates = (date: Date): boolean => {
+    return !this.disabledDates.some(({ begin, end }) => isBetween(date, begin, end));
```
Comment (pldespaigne) [inline src/app/datepicker/datepicker.component.ts]: `return !this.disabledDates.some(range => isBetween(date, range.begin, range.end));`

File: src/app/datepicker/datepicker.component.ts [hunk 22]
```diff
@@ -14,19 +18,69 @@ export class DatepickerComponent {
...
       }
+      this.validateDatepickerInput(this.datepickerRange);
     }
-    return true;
   }
+
+  validateDatepickerInput(datepickerRange: CalendarRange){
+    const { begin, end } = datepickerRange;    
+    const isInvalid = (disabledDate: CalendarRange) => {
+      return isBetween(disabledDate.begin, begin, end) || isBetween(disabledDate.end, begin, end);
```
Comment (pldespaigne) [inline src/app/datepicker/datepicker.component.ts]: `return isBetween(disabledDate.begin, datepickerRange.begin, datepickerRange.end) || isBetween(disabledDate.end, datepickerRange.begin, datepickerRange.end);`

File: src/app/month-calendar/month-calendar.component.ts [hunk 23]
```diff
@@ -19,15 +19,17 @@ export class CalendarComponent implements OnInit {
...
   months: Date[];
   currentDate: Date;
   monthTiles: Tile[];
   headerTile: Tile;  
   displayedDates: Date[];
-  startRange: Date;
-  endRange: Date;
+  begin: Date;
+  end: Date;
+  public isClicked = [];
```
Comment (pldespaigne) [inline src/app/month-calendar/month-calendar.component.ts]: `public` is by default, either remove it or put it every where

File: src/app/month-calendar/month-calendar.component.ts [hunk 24]
```diff
@@ -50,18 +52,18 @@ export class CalendarComponent implements OnInit {
...
-    const startRange = new Date(this.currentDate.getFullYear(), month, 1);
-    const endRange = new Date(this.currentDate.getFullYear(), month + 1, 0);
-    this.startRange =  startRange;
-    this.endRange = endRange;
-    return {startRange, endRange};
+    const begin = new Date(this.currentDate.getFullYear(), month, 1);
+    const end = new Date(this.currentDate.getFullYear(), month + 1, 0);
+    this.begin =  begin;
+    this.end = end;
+    return {begin, end};
```
Comment (pldespaigne) [inline src/app/month-calendar/month-calendar.component.ts]: factorize into
```typescript
this.begin = new Date(this.currentDate.getFullYear(), month, 1);
this.end = new Date(this.currentDate.getFullYear(), month + 1, 0);
return {this.begin, this.end};
```

**Pull request merged**